refactor event schedulers to use explicit game rules

IProduceWidgets opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->

Refactors basic & ramping event scheduler to take explicit event definitions in yml instead of just doing whatever the hell they wanted to. Also ported meteor scheduler over to the other scheduler types with more variety.

Added two (really just tweaked) gamemodes as rare chances in secret to flex the scheduler utility a bit.

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->

Shuttle events, gift events, the desire to add things that shouldn't be in event schedulers.
I cannot overstate how many levers this adds to event balancing by using combinations of multiple schedulers, table selectors & station event weights.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->

Now uses EntityTableSelectors. Also, since it allows for multiple uses of the different event schedulers I removed the ccvars for meteors and ramping schedulers and placed them on the components.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->
The cvars for ramping event scheduler severity and for meteor timing have been removed, and moved to fields on their respective components.
If you have events you want run via the basic/ramping event scheduler, they need to be defined explicitly now.

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
- tweak: shuttle events have been separated from other midround events and can now happen concurrently.
- tweak: Cargo Gift events should be less likely.
- tweak: meteors have been changed to be more responsive to player count.
- add: Game mode: Kessler syndrome - What if survival but also the apocalypse?
- add: Game mode: Zombeteors - Zombies and a meteor shower all at once!

TheShuEd commented:
would it make sense to exclude rules from a possible choice if it has already been selected before?  thus ensuring that all rules are selected before they start repeating

IProduceWidgets commented:
> would it make sense to exclude rules from a possible choice if it has already been selected before? thus ensuring that all rules are selected before they start repeating

You could achieve this by rule by setting the specific rule's [params](https://github.com/space-wizards/space-station-14/blob/f041f58a6dc172992e75569ca07ca7c9877590e7/Content.Server/StationEvents/Components/StationEventComponent.cs#L76C1-L76C32)

IProduceWidgets commented:
Survival not getting gifts from NT was an accident, but it is funny, so I'm leaving it.

IProduceWidgets commented:
Ready for review. I handled all the issues I could dream up.

metalgearsloth commented on Content.Server/GameTicking/Rules/Components/RandomRuleComponent.cs:
```diff
@@ -0,0 +1,25 @@
+using Content.Shared.Storage;
+
+namespace Content.Server.GameTicking.Rules.Components;
+
+[RegisterComponent, Access(typeof(RandomRuleSystem))]
+public sealed partial class RandomRuleComponent : Component
+{
```
Needs xmldoc.

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,79 @@
+using Content.Server.StationEvents.Components;
+
+namespace Content.Server.GameTicking.Rules;
+
+public sealed class RandomRuleSystem : GameRuleSystem<RandomRuleComponent>
+{
```
Same here

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,79 @@
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly IAdminLogManager _adminLogger = default!;
+    [Dependency] private readonly IChatManager _chatManager = default!;
+    [Dependency] private readonly IComponentFactory _compFact = default!;
```
This is already on entitymanager.

IProduceWidgets replied:
I don't think(?) I can use that since I need to pass ICompFac to TryGetComp since i only have a prototype and not an EntID since its not spawned yet.

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,79 @@
+    [Dependency] private readonly IChatManager _chatManager = default!;
+    [Dependency] private readonly IComponentFactory _compFact = default!;
+    [Dependency] private readonly EventManagerSystem _event = default!;
+
+    private string _ruleCompName = default!;
+    public override void Initialize()
+    {
+        base.Initialize();
+    }
+
```
Dummy component also field needs doc.

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    protected override void Added(EntityUid uid, RandomRuleComponent component, GameRuleComponent gameRule, GameRuleAddedEvent args)
+    {
+        var selectedRules = EntitySpawnCollection.GetSpawns(component.SelectableGameRules, _random);
+
+        Random rnd = new Random();
+        int ruleQuant = rnd.Next(component.MinRules, component.MaxRules);
+
```
Dummy random.

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,79 @@
+            var availableEvents = _event.AvailableEvents(); // handles the player counts and individual event restrictions, we need to do it each i incase it changes
+            string? slag = null;
+
+            foreach (var rule in selectedRules)
+            {
+                if (selectedRules == null)
+                    return;
```
If this is nullable it should be returning above (but I'm pretty sure it's not) instead of inside the loop where it's most expensive.

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,79 @@
+                if (selectedRules == null)
+                    return;
+                // If the station is already initialized, just start the rule, otherwise let that happen at the start of round.
+                if (GameTicker.RunLevel <= GameRunLevel.InRound)
+                {
+                    GameTicker.AddGameRule(rule);
+                }
+                else
+                {
+                    _prototypeManager.TryIndex(rule, out var ruleEnt);
+                    if (ruleEnt == null)
+                    {
```
Just inline tryindex.

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,79 @@
+                }
+                slag = rule;
+            }
+
+            // prevents same gamerule twice in one run.
+            if (!string.IsNullOrEmpty(slag))
+            {
+                selectedRules.RemoveAt(selectedRules.IndexOf(slag));
+            }
```
This is incredibly expensive and looks like it shouldn't be here.

IProduceWidgets replied:
I changed it to shuffling the list and just grabbing the first N elements

metalgearsloth commented on Content.Server/StationEvents/BasicStationEventSchedulerSystem.cs:
```diff
@@ -23,8 +23,9 @@ public sealed class BasicStationEventSchedulerSystem : GameRuleSystem<BasicStati
 
-        public const float MinEventTime = 60 * 3;
-        public const float MaxEventTime = 60 * 10;
+        // someome yell at me if I forget to change this back, I need to test it without dying of old
+        public const float MinEventTime = 60 * 3; // 60 * 3;
+        public const float MaxEventTime = 60 * 10; // 60 * 10;
```
Please change these back.

metalgearsloth commented on Content.Server/StationEvents/Components/RampingStationEventSchedulerComponent.cs:
```diff
@@ -14,4 +16,10 @@ public sealed partial class RampingStationEventSchedulerComponent : Component
 
     [DataField("timeUntilNextEvent"), ViewVariables(VVAccess.ReadWrite)]
     public float TimeUntilNextEvent;
+
+    /// <summary>
+    /// The gamerules that the scheduler can choose from
+    /// </summary>
+    [DataField(required: true)]
+    public List<EntitySpawnEntry> ScheduledGameRules = new();
 }
```
Aren't these supposed to be components on the entity anyway.

IProduceWidgets replied:
Some use type to subset the events like the secret gamerule, but that doesn't allow for using them in multiple different places with different balance if you wanted to. I didn't remove the type schema though because I didn't want to mess up the secret game rule or something else I hadn't considered.

metalgearsloth commented on Content.Server/StationEvents/EventManagerSystem.cs:
```diff
@@ -7,11 +8,13 @@
 
 public sealed class EventManagerSystem : EntitySystem
 {
+    [Dependency] private readonly IComponentFactory _compFac = default!;
     [Dependency] private readonly IConfigurationManager _configurationManager = default!;
```
As above.

metalgearsloth commented on Content.Server/StationEvents/EventManagerSystem.cs:
```diff
@@ -60,6 +63,71 @@ public string RunRandomEvent()
         return FindEvent(availableEvents);
     }
 
+    /// <summary>
+    /// Randomly runs an event from provided EntitySpawnCollection.
+    /// </summary>
+    public string? RunLimitedEvent(List<EntitySpawnEntry>? limitedEventsList)
+    {
+        if (limitedEventsList != null)
+        {
+            var availableEvents = AvailableEvents(); // handles the player counts and individual event restrictions
+
+            if (availableEvents.Count == 0)
+            {
+                Log.Warning("No events were available to run!");
+                return null;
+            }
```
Having nullable APIs like this is bad, just add overloads.

IProduceWidgets replied:
ah smort. I always forget about overloads.

IProduceWidgets commented:
Should all be addressed with the exception of the IComponentFactory dependency thing which I think I need a hint on if its a big issue.

EmoGarbage404 commented:
> with the exception of the IComponentFactory dependency thing which I think I need a hint on if its a big issue.

you can grab it by doing `EntityManager.ComponentFactory`

IProduceWidgets commented:
> you can grab it by doing `EntityManager.ComponentFactory`

Thanks! done!

EmoGarbage404 commented on Content.Server/StationEvents/Components/BasicStationEventSchedulerComponent.cs:
```diff
@@ -1,14 +1,22 @@
 public sealed partial class BasicStationEventSchedulerComponent : Component
 {
-    public const float MinimumTimeUntilFirstEvent = 300;
+    public const float MinimumTimeUntilFirstEvent = 300; // Someone yell at me if I dont change this back to 300, I need to test it without dying of old
```
```suggestion
    public const float MinimumTimeUntilFirstEvent = 300
```

TheShuEd replied:
Noo! Dont remove ;

EmoGarbage404 commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,80 @@
+    [Dependency] private readonly IChatManager _chatManager = default!;
+    [Dependency] private readonly EventManagerSystem _event = default!;
+
+    protected override void Added(EntityUid uid, RandomRuleComponent component, GameRuleComponent gameRule, GameRuleAddedEvent args)
```
this should probably happen on `Started` instead of `Added`

IProduceWidgets replied:
See last review for why Added may be better.

EmoGarbage404 commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,80 @@
+    {
+        var selectedRules = EntitySpawnCollection.GetSpawns(component.SelectableGameRules, _random);
+
+        if (component.MinRules > component.MaxRules)
+        {
+            Log.Warning("Minimum rules is more than Maximum rules!");
+            return;
+        }
+
+        int ruleQuant = _random.Next(component.MinRules, component.MaxRules + 1); // the padding is required for expected result. Dont ask me why Next was implimented this way.
```
you could just a MinMax struct instead of two variables and then there's a helper function for Next.

also it was implemented that way because that's how it works on basically everything.

IProduceWidgets replied:
Oh nice, thanks I'll use the MinMax.

Unfortunately the helper function exhibits the same issue I had to fix with the padding. Not doing that means you need to define a Max of 3 to get 2 things which is dumb as fuck. It should just be changed to not be stupid but that's way beyond this prs purview.

EmoGarbage404 commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,80 @@
+                    continue;
+                }
+
+                if (ruleProto.TryGetComponent<StationEventComponent>(out _, EntityManager.ComponentFactory))
+                {
+                    if (!availableEvents.ContainsKey(ruleProto))
+                    {
```
i dont see why this check is happening. this is just hardcoding this system to only work with station events when in reality it can work with whatever the hell it wants.

EmoGarbage404 commented on Resources/Prototypes/GameRules/cargo_gifts.yml:
```diff
@@ -1,3 +1,42 @@
+  components:
+  - type: RandomRule
+    selectableGameRules:
+    # Make the probability sum to 1 or I will remove your fingernails with toothpicks.
```
you dont need to make it sum to one. this uses weights for a reason.

IProduceWidgets replied:
Need is relative, balancing 85 gamerules with random weights sounds like cbt to me, but if you guys would rather keep track of that than the clear and concise "it happens 15% of the time" its no skin off my back.

metalgearsloth replied:
All the probabilities are the same you may as well remove the yaml lines and make it significantly easier to maintain.

EmoGarbage404 commented on Resources/Prototypes/GameRules/roundstart.yml:
```diff
@@ -254,12 +254,73 @@
+    - id: MimicVendorRule
+    - id: LoneOpsSpawn
+    - id: RevenantSpawn
+#    - id: MassHallucination    # What the fuck is this? Is it its own whoosiwhatsit and not an event?
```
its a station event and idk why you commented it out

IProduceWidgets replied:
Oh I see what happened. I skill issued and tried to add the tracking component and not the rule component. Easy fix.

EmoGarbage404 commented on Resources/Prototypes/GameRules/roundstart.yml:
```diff
@@ -254,12 +254,73 @@
+    - id: SolarFlare
+    - id: KudzuGrowth
+    - id: ZombieOutbreak
+#    - id: CargoGiftMasterRule
```
same here?

IProduceWidgets replied:
This one was just an accident, but I thought it was funny that survival doesn't get the mercy gifts.

EmoGarbage404 commented on Resources/Prototypes/GameRules/unknown_shuttles.yml:
```diff
@@ -1,3 +1,25 @@
+  components:
+  - type: RandomRule
+    selectableGameRules:
+    # Make the probability sum to 1 or I will remove your fingernails with toothpicks.
```
same thing here. there's no point in requiring it to sum to one and doing that will just lead to breaking changes on this file for no reason

metalgearsloth replied:
Remove the 0.25 probs and make the 0.3 prob 1.2.

EmoGarbage404 commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            var ruleEnt = new EntityUid();
+
+            // If the station is already initialized, just start the rule, otherwise let that happen at the start of round.
+            if (GameTicker.RunLevel <= GameRunLevel.InRound)
+            {
+                ruleEnt = GameTicker.AddGameRule(rule);
+            }
```
does this even make sense? ik antags have this behavior but they actually have code that executes on roundstart. if you add a station event pre-roundstart and never start it manually it just won't do anything.

IProduceWidgets replied:
Ahh, okay, yeah. This is the reason I used Added. You could potentially use the RandomGameRule to add roundstart events if say you wanted a secret gamemode that chose a random number of round start events.

I don't necessarily advocate doing that, but we lose nothing making it possible so...

IProduceWidgets replied:
Okay yeah, I thought I could use this to make it possible for this to also pick roundstart gamerules, but then it doesn't play well with stationevents, and hey are more important and we a;ready have the secretrulesystem which is good enough for roundstart rules imo.

IProduceWidgets commented:
One sec I noticed an issue

IProduceWidgets commented:
I fixed the issues.

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,61 @@
+
+        var availableEvents = _event.AvailableEvents(); // handles the player counts and individual event restrictions, we need to do it each i incase it changes
+
+        foreach (var rule in nRules)
+        {
+            var ruleEnt = new EntityUid();
+
```
You shouldn't need to construct this ruleent here and can remove it.

metalgearsloth commented on Content.Server/GameTicking/Rules/RandomRuleSystem.cs:
```diff
@@ -0,0 +1,61 @@
+                Log.Warning("The selected random rule is not available!");
+                continue;
+            }
+            GameTicker.StartGameRule(rule, out ruleEnt);
+
+            var str = Loc.GetString("station-event-system-run-event", ("eventName", ToPrettyString(ruleEnt)));
+            _chatManager.SendAdminAlert(str);
+            Log.Info(str);
+        }
```
Isn't this already admin logged, hardcoding in logs for 1 rule looks like a workaround.

IProduceWidgets replied:
Fixed it, but it does do this at round init now, and I don't see an easy way around that without special casing again.
![image](https://github.com/user-attachments/assets/c4726a31-76c9-4a7e-8f41-6f9086256743)

IProduceWidgets replied:
Maybe I could suppress it before round start? 🤔

metalgearsloth commented on Content.Server/StationEvents/EventManagerSystem.cs:
```diff
@@ -44,12 +46,77 @@ public string RunRandomEvent()
         }
 
         var ent = GameTicker.AddGameRule(randomEvent);
-        var str = Loc.GetString("station-event-system-run-event",("eventName", ToPrettyString(ent)));
+        var str = Loc.GetString("station-event-system-run-event", ("eventName", ToPrettyString(ent)));
         _chat.SendAdminAlert(str);
         Log.Info(str);
         return str;
     }
```
As above just copy-pasting this is bad and needs fixing.

metalgearsloth commented on Content.Server/StationEvents/EventManagerSystem.cs:
```diff
@@ -44,12 +46,77 @@ public string RunRandomEvent()
+                Log.Warning("A requested event is not available!");
+                return null;
+            }
+
+            var ent = GameTicker.AddGameRule(randomLimitedEvent);
+            var str = Loc.GetString("station-event-system-run-event", ("eventName", ToPrettyString(ent)));
+            _chat.SendAdminAlert(str);
+            Log.Info(str);
+            return str;
+        }
```
As above.

metalgearsloth commented on Content.Server/StationEvents/EventManagerSystem.cs:
```diff
@@ -44,12 +46,77 @@ public string RunRandomEvent()
+            Log.Info(str);
+            return str;
+        }
+        Log.Warning("A scheduler has requested null as event!");
+        return null;
+    }
+
```
Why does this even allow null inputs? Just make the method non-nullable instead of just logging a needless warning.

IProduceWidgets replied:
Fixed. I changed the methods to void, since they were just returning a string, but as far as I can tell, nothing else interacts with that string anywhere.

metalgearsloth commented on Resources/Prototypes/GameRules/roundstart.yml:
```diff
@@ -254,12 +254,74 @@
   parent: BaseGameRule
   components:
   - type: RampingStationEventScheduler
+    scheduledGameRules:
+    - id: ClosetSkeleton
+    - id: MouseMigration
+    - id: SlimesSpawn
+    - id: BluespaceArtifact
+    - id: ClericalError
+    - id: IonStorm
+    - id: KingRatMigration
+    - id: CockroachMigration
+    - id: SpiderClownSpawn
+    - id: VentClog
+    - id: GasLeak
+    - id: ImmovableRodSpawn
+    - id: BreakerFlip
+    - id: PowerGridCheck
+    - id: BluespaceLocker
+    - id: NinjaSpawn
+    - id: BureaucraticError
+    - id: DragonSpawn
+    - id: SpiderSpawn
+    - id: MimicVendorRule
+    - id: LoneOpsSpawn
+    - id: RevenantSpawn
+    - id: MassHallucinations
+    - id: SleeperAgents
+    - id: AnomalySpawn
+    - id: SolarFlare
+    - id: KudzuGrowth
+    - id: ZombieOutbreak
+    - id: RandomCargoGiftsRule
+    - id: RandomUnknownShuttleRule
```
Please just alphabetise it this is unmaintainable.

metalgearsloth commented on Resources/Prototypes/GameRules/roundstart.yml:
```diff
@@ -254,12 +254,74 @@
   parent: BaseGameRule
   components:
   - type: BasicStationEventScheduler
+    scheduledGameRules:
+    - id: ClosetSkeleton
+    - id: MouseMigration
+    - id: SlimesSpawn
+    - id: BluespaceArtifact
+    - id: ClericalError
+    - id: IonStorm
+    - id: KingRatMigration
+    - id: CockroachMigration
+    - id: SpiderClownSpawn
+    - id: VentClog
+    - id: GasLeak
+    - id: ImmovableRodSpawn
+    - id: BreakerFlip
+    - id: PowerGridCheck
+    - id: BluespaceLocker
+    - id: NinjaSpawn
+    - id: BureaucraticError
+    - id: DragonSpawn
+    - id: SpiderSpawn
+    - id: MimicVendorRule
+    - id: LoneOpsSpawn
+    - id: RevenantSpawn
+    - id: MassHallucinations
+    - id: SleeperAgents
+    - id: AnomalySpawn
+    - id: SolarFlare
+    - id: KudzuGrowth
+    - id: ZombieOutbreak
+    - id: RandomCargoGiftsRule
+    - id: RandomUnknownShuttleRule
 
```
Same here.

IProduceWidgets commented:
Guh, what did I break

IProduceWidgets commented:
Woohoo She floats!

Okay I updated this to use EntitySpawnEntries.

The only thing I'm kind of uncertain about working is the toolshed commands in the BasicStationEventSchedulerSystem.
I have no idea how to test them, so I just updated the logic to how they should work with the new stuff. I don't know if its okay to make them take an EntityPrototype as an argument, maybe it needs to be a string and then cast. If someone can make me a toolshed call that works on the current build I'll make sure it works. I've never seen any of those commands get used before though, and I have nfi how to use them.

kanopus952 commented:
Hello! Do you know of a bug that may have been caused by your changes? Check this issue: https://github.com/space-wizards/space-station-14/issues/31090

IProduceWidgets commented:
> Hello! Do you know of a bug that may have been caused by your changes? Check this issue: #31090

I have been sent a couple occurrences, but have thus far been unable to replicate it. Please do let me know of more occurrences, or if you find a consistent occurrence.
